Double ended sliders extension example

djpugh opened the pull request:
issue: #1996 - this PR adds a documentation extensions example implementing two variants of a double ended slider, which maybe a useful starting point for fixing #1996

bryevdv commented:
HI @djpugh thanks for the PR! It looks like there is one real test failure, our linter noted this:
```
E       AssertionError: Code quality issues:
E         File ends with more than 1 empty line: sphinx/source/docs/user_guide/extensions_gallery/widget.rst, line 13
E       assert 1 == 0
E        +  where 1 = len(['File ends with more than 1 empty line: sphinx/source/docs/user_guide/extensions_gallery/widget.rst, line 13'])
```
Once you delete the extra empty lines the tests should start passing.

djpugh commented:
@bryevdv - thanks, I've just done that so hopefully the tests will pass now!

mattpap commented on sphinx/source/docs/user_guide/source_examples/extensions_example_widget.py:
```diff
@@ -0,0 +1,416 @@
+
+callback_jquery = CustomJS(args=dict(source=source), code="""
+        var data = source.get('data');
+        var f = cb_obj.get('range')
```
`obj.get('attr')` is deprecated in favor of `objt.attr` (getters). This and all other instances should be change.

djpugh replied:
@mattpap - as in `cb_obj.attr('range')` etc?

bryevdv replied:
@djpugh no, previously we had `obj.get('foo')` everywhere. Now we can (and should) always have just `obj.foo`

mattpap replied:
> as in cb_obj.attr('range') etc?

@djpugh, you mean `cb_obj.get('range')`?. Yes, it will be simply `cb_obj.range`.

djpugh replied:
@mattpap, @bryevdv - thanks!

mattpap commented on sphinx/source/docs/user_guide/source_examples/extensions_example_widget.py:
```diff
@@ -0,0 +1,416 @@
+"""
+
+
+ion_range_slider_coffee = """
```
That's a lot code to be processed without syntax highlighting. Maybe this could be moved to a `*.coffee` file and `__implementation__` changed accordingly, e.g. `__implementation__ = "ion_range_slider.coffee"`.

bryevdv replied:
agreed this would make the example a bit easier to read but it will also require some manual intervention in `extensions_gallery/widget.rst` the `bokeh-plot` directive is not sophisticated enough to handle multiple source files. An option would be to not have it display the source at all, but instead put in `code-block`s that show the individual files with highlighting

djpugh replied:
My current plan was to do something like:
```
.. literalinclude::  source/docs/user_guide/source_examples/extensions_jquery_ui_range_slider.coffee
   :language: coffee
   :linenos:
```
Will bokeh plot pull in the external file correctly for generating the plot?

bryevdv replied:
@djpugh that seems like it would work great

djpugh replied:
@bryevdv - Sure, I'll test it out and see if it works!

mattpap commented on sphinx/source/docs/user_guide/source_examples/extensions_example_widget.py:
```diff
@@ -0,0 +1,416 @@
+# We will subclass in JavaScript from the same class that was subclassed
+# from in Python
+InputWidget = require "models/widgets/input_widget"
+ionslidertemplate = require "models/widgets/slidertemplate"
```
It would be better to just copy and update this template and not use `html.replace()` later on. Unfortunately, our custom module compiler doesn't allow to compile eco templates right now.

djpugh replied:
@mattpap - I had had a go with that, and couldn't get it to work, which makes sense given your comment - I've added a comment in the coffeescript to that effect.

djpugh replied:
@bryevdv - I've opened an issue #5279 for this.

mattpap replied:
@djpugh, I added support for eco templates in PR #5280.

djpugh replied:
and will it load correctly with bokeh-plot or should I just show it in the code blocks?

bryevdv replied:
ping @mattpap does the template also need to be specified as an available resource somewhere?

mattpap replied:
It just has to be `require()`d as any other module (no extension, though it may work with one).

djpugh replied:
It doesn't seem to be working with the `.eco` ending on the require statement

djpugh replied:
Debugging that error a bit it looks as if the eco template is not generating any code in `bokeh\utils\compiler.py:nodejs_compile` - I'm just going to try stepping through the python code to check nothing weird is happening there

djpugh replied:
I don't think it's anything in my template as using the stock slider template it also generates no compiled code, so will error here (Line 289-294 of `bokeh\utils\compiler.py`):
```
                if impl.lang == "less":
                    code = _style_template % dict(css=json.dumps(compiled.code))
                    deps = []
                else:
                    code = compiled.code
                    deps = compiled.deps
```

djpugh replied:
@bryevdv - it is

mattpap replied:
@djpugh, did you rebuild bokehjs? Though bokehjs and custom module compiler are independent things, you have to issue `gulp build` whenever compiler changes (as they share the same build).

djpugh replied:
Ah thanks @mattpap - thought I had this morning, but obviously didn't - have just rebuilt and it all seems to be working

mattpap commented on sphinx/source/docs/user_guide/source_examples/extensions_example_widget.py:
```diff
@@ -0,0 +1,416 @@
+    html = @template(@model.attributes)
+    # replace div slider to input slider
+    html = html.replace('<div class="slider "','<input type="text" class="slider "')
+    html = html.replace('<div class="bk-slider-horizontal">','<div class="bk-slider-horizontal bk-ion-slider">')
```
This isn't going to work if user configures `orientation` differently.

bryevdv replied:
@mattpap ion range slider does not support vertical orientations at all. @djpugh I think `orientation` should just be removed completely from the available properties

djpugh replied:
@bryevdv - Have done, just setting up the build on a different machine to test, so it may be a little while until I push the new commits

bryevdv replied:
awesome, thanks!

bryevdv commented:
@djpugh, I agree with @mattpap comments do you expect to have time to devote to this? If not, I understand. In that case it is possible to allow one of us to push directly to your PR branch, if that is acceptable collaboration.

djpugh commented:
@bryevdv, @mattpap - I've addressed those comments - I couldn't get the bokeh plot to work with using external implementation files, so as @bryevdv suggested I just created an additional file for the `bokeh-plot` command and used the separated source files.

bryevdv commented on sphinx/source/docs/user_guide/source_examples/extensions_example_widget.py:
```diff
@@ -303,7 +76,7 @@ class JQueryRangeSlider(InputWidget):
     # of the custom extension model.
 
-    __implementation__ = jquery_ui_range_slider_coffee
+    __implementation__ = open('extensions_jquery_ui_range_slider.coffee').read()
```
@djpugh this (and the one above) can actually just be the filename  `"extensions_jquery_ui_range_slider.coffee"` Any string ending in `.coffee` will automatically be treated as a file and read in. This is what we should demonstrate.

bryevdv commented:
@djpugh I just have one final comment about `__implementation__. Also if you can, a separate issue about improving `bokeh-plot` to support this kind of use case would be good to record.

djpugh commented:
@bryevdv - I've changed the `__implementation__` to the filename and added a comment in the code too.

bryevdv commented:
ok @djpugh #5280 for supporting eco templates in custom models has been merged. Do you want to try and work on updating for that in the next couple of days? If not, no worries, we will get it in for `0.12.4` then

djpugh commented:
@bryevdv - happy to add the eco template, just need to know how to include it in the CoffeeScript implementation - I've tried a couple of things and I'm getting `RuntimeError: no such module: ` errors at the moment, but if @mattpap let's me know how to do it, it shouldn't take me long to get done

mattpap commented:
@djpugh, you have merge most recent `master` branch to this branch before you can use Eco templates.

djpugh commented:
@bryevdv, @mattpap - I've just  pushed the changes with the template, it is finding it, but as mentioned above, it isn't generating any compiled code, so then errors.

bryevdv commented:
@djpugh OK, let's hold this PR for `0.12.4` then, it should be out very soon after `0.12.3` (just a couple of weeks) @mattpap can help debug whether the problem is on the example side, or the bokeh side. Thanks for your patience!

djpugh commented:
@bryevdv - sure, I've added some diagnostics back to issue #5279 - should I open another issue or leave that one?

djpugh commented:
@bryevdv - would it be useful to add the jqueryui implementation into the standard models if we are going to hold for 0.12.4?

bryevdv commented:
@djpugh It would be nice but it is too late, we have to test with real RC tomorrow to have any hope of releasing by Friday. I expect `0.12.4` will only be a few weeks later, and in any case we can put out a `0.12.4dev` build at any time easily.

djpugh commented:
@bryevdv - I meant for the 0.12.4 release as well, changing this to just show the ion range slider example

djpugh commented:
@bryevdv - following me rebuilding `bokehjs` this is now working!

djpugh commented:
However, when building the docs after a `make clean` I get an error: 

```
Sphinx error:
Unable to generate Bokeh plot at \bokeh\sphinx\source\docs\gallery\anscombe.rst:12:no such module: ./extensions_ion_range_slider_template
```

I'm guessing this is a problem with the docs build hooks for this @mattpap?

bryevdv commented:
@djpugh Yes probably. I actually want to make some improvements to `:bokeh-plot:` and this PR has several of the motivating uses. Let's definitely hold off for `0.12.3` and I will make another PR off of this branch to improve `:bokeh-plot:`

djpugh commented:
@bryevdv - sure, I'll see if i can split of the jqueryUI example into the core models as well, and make this just a demo of the ionrangeslider. 

Sphinx can generate the actual plot in question, it just fails on the examples, so I think it must be doing something odd there!

djpugh commented:
I've updated this to just use the ionRangeSlider assuming that we can make #5309 acceptable for merging too.

bryevdv commented:
@djpugh absolutely, I'd be OK merging this tonight before building `0.12.3` tomorrow, and then we'd leave #5309 for `0.12.4` (which should follow on very soon).

Can you attach a screen shot of the docs pages for this? That would save me the trouble of fetching the branch and building and checking.

djpugh commented:
@bryevdv Sure, I'm just getting it built from the docs, but I'll attach when it has

djpugh commented:
I'm still getting that build error on the anscombe example when I build the docs from scratch

djpugh commented:
Docs page generated by building master then on the PR branch.

![image](https://cloud.githubusercontent.com/assets/6003255/19169746/ab469164-8c0d-11e6-9479-6f238c5cf78f.png)
![image](https://cloud.githubusercontent.com/assets/6003255/19169771/bba1c4ac-8c0d-11e6-890e-3efa8fd8d434.png)
![image](https://cloud.githubusercontent.com/assets/6003255/19169789/c65bb5d8-8c0d-11e6-9700-a6470e884c93.png)
![image](https://cloud.githubusercontent.com/assets/6003255/19169795/d110004c-8c0d-11e6-8938-ad7e2450eee2.png)
![image](https://cloud.githubusercontent.com/assets/6003255/19169813/df3c0530-8c0d-11e6-9843-f6f37d5c21db.png)
![image](https://cloud.githubusercontent.com/assets/6003255/19169823/e733e168-8c0d-11e6-8662-4569f461f8f5.png)

djpugh commented:
but it seems to have built ok on Travis

djpugh commented:
@bryevdv - feel free to edit anything if required, or let me know if anything else needs doing!

bryevdv commented:
@djpugh I built it just to see if I could repro the anscombe issue but everything built great and works great. Will merge now. Thanks for the PR!

tommycarstensen commented:
Is this currently the easiest way to do a double ended slider? Thanks.
https://docs.bokeh.org/en/latest/docs/user_guide/extensions_gallery/widget.html
